### PR TITLE
Refactors rate_limit plugin with YAML configurations

### PR DESCRIPTION
**Pull request by zwoop**
Besides majorly refactoring a lot of the global plugin code, this PR makes the following changes:

* YAML configuration
* Reloadable configurations (even as a global)
* SNI aliases (one rate limiter can apply to many SNIs)
* The IP reputation objects are now shareable for many SNIs. This means an attacker can circumvent the reputation by spreading the DDoS traffic across many SNIs.

### CODE REVIEW

**Comment by shukitchan** [comment]
I can review this when it is ready

**Comment by zwoop** [comment]
> I can review this when it is ready

Great! Almost there :). The refactoring focuses entirely on the global plugin only, leaving the remap plugin as-is. One hope with this refactoring and change to a reloadable YAML configuration is that we can add more features and rate limiting rules more easily.

**Comment by shukitchan** [comment]
@zwoop let me know when it is ready to review. It looks like it is but I am not sure if you are having more changes in your mind.

**Comment by bneradt** [comment]
[approve ci autest]

File: plugins/experimental/rate_limit/sni_selector.cc [hunk 1]
```diff
@@ -16,117 +16,240 @@
...
-sni_queue_cont(TSCont cont, TSEvent event, void *edata)
+bool
+SniSelector::yamlParser(const std::string &yaml_file)
 {
-  SniSelector *selector = static_cast<SniSelector *>(TSContDataGet(cont));
+  YAML::Node config;
+
+  try {
+    config = YAML::LoadFile(yaml_file);
+  } catch (YAML::BadFile &e) {
```
Comment (brbzull0) [inline plugins/experimental/rate_limit/sni_selector.cc:35]: You are not going to change the exception right? if not then why not `const`.
the one below too.

Comment (zwoop) [reply plugins/experimental/rate_limit/sni_selector.cc:35]: I just copied this pattern from other places in the code :). I can change it.

File: plugins/experimental/rate_limit/sni_selector.cc [hunk 2]
```diff
@@ -139,6 +262,26 @@ SniSelector::setupQueueCont()
...
+  TSReleaseAssert(sni_cont);
+
+  _instance.store(new SniSelector());
+  TSHttpHookAdd(TS_SSL_CLIENT_HELLO_HOOK, sni_cont);
+  TSHttpHookAdd(TS_VCONN_CLOSE_HOOK, sni_cont);
+
+  auto config_cont = TSContCreate(sni_config_cont, TSMutexCreate());
+
+  TSReleaseAssert(config_cont);
+  TSContScheduleEveryOnPool(config_cont, std::chrono::milliseconds{10000}.count(), TS_THREAD_POOL_TASK);
```
Comment (shukitchan) [inline plugins/experimental/rate_limit/sni_selector.cc:286]: Is there a reason why we want to change the config file every 10 seconds vs using TSMgmtUpdateRegister() and wait for an explicit config reload 

just to make it more convenient?

Can a race condition happen when the config yaml is only partial updated when the config_cont is scheduled to run and therefore resulting in some inconsistent behavior?

File: plugins/experimental/rate_limit/sni_selector.cc (hunk 1 above)
Comment (brbzull0) [reply plugins/experimental/rate_limit/sni_selector.cc]: hope it wasn't my code! that would be a shame :rofl:

File: plugins/experimental/rate_limit/sni_selector.cc (hunk 2 above)
Comment (zwoop) [reply plugins/experimental/rate_limit/sni_selector.cc:286]: Yeh, for convenience, but we can change to be triggered with traffic_ctl.

Comment (shukitchan) [reply plugins/experimental/rate_limit/sni_selector.cc:286]: I think it is better to be triggered by traffic_ctl

**Review by shukitchan — DISMISSED** [review]
I think it looks ok.

**Comment by zwoop** [comment]
[approve ci]

File: doc/admin-guide/plugins/rate_limit.en.rst [hunk 3]
```diff
@@ -122,51 +121,117 @@ The basic use is as::
...
 
-.. option:: --tag
+   The queue option can include a `size` and a `max-age` option. The size is
+   default to ``UINT_MAX``, which is essentially unlimited. The max-age is
+   default to ``0``, which means no age limit.
 
-   An optional metric tag to use instead of the default. When a tag is not specified
-   the plugin will use the FQDN of the SNI associated with each rate limiter instance
-   created during plugin initialization.
+   No queue is enable without this configuration directive, but it can also be
```
Comment (shukitchan) [inline doc/admin-guide/plugins/rate_limit.en.rst:196]: enabled?

File: plugins/experimental/rate_limit/CMakeLists.txt [hunk 4]
```diff
@@ -15,6 +15,8 @@
 #
 #######################
 
+project(rate_limit)
+
 add_atsplugin(
   rate_limit
   ip_reputation.cc
```
Comment (shukitchan) [inline plugins/experimental/rate_limit/CMakeLists.txt:22]: lists.cc is not needed herer?

**Review by shukitchan — CHANGES_REQUESTED** [review]

File: plugins/experimental/rate_limit/CMakeLists.txt (hunk 4 above)
Comment (zwoop) [reply plugins/experimental/rate_limit/CMakeLists.txt:22]: Not sure why it didn't fail without lists.cc ... Added.

**Comment by zwoop** [comment]
I updated this to use the updated version of the registration API.

File: doc/admin-guide/plugins/rate_limit.en.rst (hunk 3 above)
Comment (zwoop) [reply doc/admin-guide/plugins/rate_limit.en.rst]: fixed

**Pull request merged**